File: src/input/mouse.ts
// SGR mouse event parser and encoder for terminal mouse tracking
// SGR format: \x1b[<Cb;Cx;CyM (press) or \x1b[<Cb;Cx;Cym (release)
// Coords are 1-based in the protocol, converted to 0-based in MouseEvent

export interface MouseEvent {
  button: number; // raw Cb value (includes modifiers + motion flag)
  x: number; // 0-based column
  y: number; // 0-based row
  isRelease: boolean;
}

export const MOUSE_LEFT = 0;
export const MOUSE_MIDDLE = 1;
export const MOUSE_RIGHT = 2;
export const MOUSE_SCROLL_UP = 64;
export const MOUSE_SCROLL_DOWN = 65;

/** Extract base button (0=left, 1=middle, 2=right, 64+=scroll) */
export function getBaseButton(button: number): number {
  // Bits 0-1 encode button, bit 6 is scroll flag
  // Mask out modifier bits (4=shift, 8=meta, 16=ctrl) and motion bit (32)
  return button & 0b1000011;
}

/** Check if this is a scroll event */
export function isScrollEvent(button: number): boolean {
  return (button & 64) !== 0;
}

/** Check if this is a motion event (mouse drag / mouse move) */
export function isMotionEvent(button: number): boolean {
  return (button & 32) !== 0;
}

export interface ParseResult {
  event: MouseEvent;
  consumed: number; // number of bytes consumed from the buffer
}

/**
 * Parse an SGR mouse sequence from a buffer.
 * Returns null if the buffer doesn't start with a valid SGR mouse sequence.
 * Handles partial sequences by returning null (caller should buffer and retry).
 */
export function parseSgrMouse(data: Buffer, offset = 0): ParseResult | null {
  // Minimum: \x1b[<0;1;1M = 9 bytes
  if (offset + 9 > data.length) return null;
  if (
    data[offset] !== 0x1b ||
    data[offset + 1] !== 0x5b ||
    data[offset + 2] !== 0x3c
  ) {
    return null;
  }

  // Find the terminator (M or m)
  let end = offset + 3;
  while (end < data.length) {
    const byte = data[end]!;
    if (byte === 0x4d || byte === 0x6d) {
      // 'M' or 'm'
      break;
    }
    // Valid chars in the sequence: digits and semicolons
    if ((byte >= 0x30 && byte <= 0x39) || byte === 0x3b) {
      end++;
      continue;
    }
    // Invalid character — not a mouse sequence
    return null;
  }

  if (end >= data.length) {
    // Incomplete sequence
    return null;
  }

  const isRelease = data[end] === 0x6d; // 'm' = release, 'M' = press
  const params = data.subarray(offset + 3, end).toString("ascii");
  const parts = params.split(";");

  if (parts.length !== 3) return null;

  const button = parseInt(parts[0]!, 10);
  const x = parseInt(parts[1]!, 10) - 1; // Convert to 0-based
  const y = parseInt(parts[2]!, 10) - 1; // Convert to 0-based

  if (isNaN(button) || isNaN(x) || isNaN(y)) return null;

  return {
    event: { button, x, y, isRelease },
    consumed: end - offset + 1,
  };
}

/**
 * Encode a mouse event back to SGR format with new (pane-local) coordinates.
 * Coordinates are 0-based, will be converted to 1-based for the protocol.
 */
export function encodeSgrMouse(
  button: number,
  localX: number,
  localY: number,
  isRelease: boolean,
): string {
  const terminator = isRelease ? "m" : "M";
  return `\x1b[<${button};${localX + 1};${localY + 1}${terminator}`;
}

File: rust/crates/maxmux-input/src/mouse.rs
#[derive(Debug, Clone, PartialEq)]
pub struct MouseEvent {
    pub button: u8,
    pub x: u16, // 0-based
    pub y: u16, // 0-based
    pub is_release: bool,
}

pub const MOUSE_LEFT: u8 = 0;
pub const MOUSE_MIDDLE: u8 = 1;
pub const MOUSE_RIGHT: u8 = 2;
pub const SCROLL_UP: u8 = 64;
pub const SCROLL_DOWN: u8 = 65;

/// Parse SGR mouse event from bytes.
/// Format: ESC[<Cb;Cx;CyM (press) or ESC[<Cb;Cx;Cym (release)
/// Returns (MouseEvent, total bytes consumed INCLUDING the ESC[< prefix)
pub fn parse_sgr_mouse(data: &[u8]) -> Option<(MouseEvent, usize)> {
    // Look for ESC[< prefix
    if data.len() < 3 || data[0] != 0x1b || data[1] != b'[' || data[2] != b'<' {
        return None;
    }

    let rest = &data[3..];

    // Parse Cb;Cx;Cy terminated by M or m
    // Find terminator M or m
    let term_pos = rest.iter().position(|&b| b == b'M' || b == b'm')?;

    let params_str = std::str::from_utf8(&rest[..term_pos]).ok()?;
    let parts: Vec<&str> = params_str.split(';').collect();
    if parts.len() != 3 {
        return None;
    }

    let button: u8 = parts[0].parse().ok()?;
    let x: u16 = parts[1].parse().ok()?;
    let y: u16 = parts[2].parse().ok()?;
    let is_release = rest[term_pos] == b'm';

    // Coords are 1-based in protocol, convert to 0-based
    let event = MouseEvent {
        button,
        x: x.saturating_sub(1),
        y: y.saturating_sub(1),
        is_release,
    };

    // Total consumed: 3 (ESC[<) + params + 1 (M or m)
    let consumed = 3 + term_pos + 1;
    Some((event, consumed))
}

/// Encode a mouse event to SGR format.
/// Coords are 0-based, will be converted to 1-based for the protocol.
pub fn encode_sgr_mouse(button: u8, x: u16, y: u16, is_release: bool) -> String {
    format!(
        "\x1b[<{};{};{}{}",
        button,
        x + 1,
        y + 1,
        if is_release { 'm' } else { 'M' }
    )
}

/// Extract the base button number (0=left, 1=middle, 2=right).
pub fn base_button(button: u8) -> u8 {
    button & 0b11
}

/// Check if this is a scroll event.
pub fn is_scroll(button: u8) -> bool {
    button & 64 != 0
}

/// Check if this is a motion event.
pub fn is_motion(button: u8) -> bool {
    button & 32 != 0
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_left_click() {
        let data = b"\x1b[<0;10;5M";
        let (event, consumed) = parse_sgr_mouse(data).unwrap();
        assert_eq!(event.button, 0);
        assert_eq!(event.x, 9); // 10-1
        assert_eq!(event.y, 4); // 5-1
        assert!(!event.is_release);
        assert_eq!(consumed, data.len());
    }

    #[test]
    fn test_parse_release() {
        let (event, _) = parse_sgr_mouse(b"\x1b[<0;10;5m").unwrap();
        assert!(event.is_release);
    }

    #[test]
    fn test_encode_roundtrip() {
        let encoded = encode_sgr_mouse(0, 9, 4, false);
        assert_eq!(encoded, "\x1b[<0;10;5M");
    }

    #[test]
    fn test_scroll_detection() {
        assert!(is_scroll(64));
        assert!(!is_scroll(0));
    }

    #[test]
    fn test_base_button() {
        assert_eq!(base_button(MOUSE_LEFT), 0);
        assert_eq!(base_button(MOUSE_RIGHT), 2);
        assert_eq!(base_button(MOUSE_MIDDLE), 1);
    }

    #[test]
    fn test_motion_detection() {
        assert!(is_motion(32));
        assert!(!is_motion(0));
    }

    #[test]
    fn test_parse_invalid_prefix() {
        assert!(parse_sgr_mouse(b"hello").is_none());
    }

    #[test]
    fn test_parse_incomplete() {
        assert!(parse_sgr_mouse(b"\x1b[<0;10;5").is_none());
    }

    #[test]
    fn test_encode_release() {
        let encoded = encode_sgr_mouse(0, 9, 4, true);
        assert_eq!(encoded, "\x1b[<0;10;5m");
    }

    #[test]
    fn test_parse_right_click() {
        let data = b"\x1b[<2;1;1M";
        let (event, consumed) = parse_sgr_mouse(data).unwrap();
        assert_eq!(event.button, 2);
        assert_eq!(event.x, 0);
        assert_eq!(event.y, 0);
        assert!(!event.is_release);
        assert_eq!(consumed, data.len());
    }
}
